pilot/pkg/model/push_context.go

```diff
@@ -2173,17 +2174,11 @@ func (ps *PushContext) ServiceAccounts(hostname host.Name, namespace string, por
 	}]
 }
 
-func (ps *PushContext) SupportsTunnel(ip string) bool {
-	infos, _ := ps.ambientIndex.AddressInformation(sets.New(types.NamespacedName{Name: ip}))
-	for _, p := range infos {
-		switch addr := p.Address.Type.(type) {
-		case *workloadapi.Address_Workload:
-			if addr.Workload.TunnelProtocol == workloadapi.TunnelProtocol_HBONE {
-				return true
-			}
-		case *workloadapi.Address_Service:
-			// Services do not directly support tunneling, their individual Workloads can though
-			continue
+func (ps *PushContext) SupportsTunnel(n network.ID, ip string) bool {
+	infos, _ := ps.ambientIndex.AddressInformation(sets.New(n.String() + "/" + ip))
+	for _, wl := range ExtractWorkloadsFromAddresses(infos) {
+		if wl.TunnelProtocol == workloadapi.TunnelProtocol_HBONE {
```

Comment:
Not sure I understand this logic - do you mind adding a small comment on the method ? 

If any of the workloads for the IP supports HBONE return true ? What about the others ? And if it's a service, what does it mean ?

Reply:
It should only actually return 1 workload. Maybe should be refactored (or commented) to be more clear

Reply:
yes, returning >1 workload for an IP is also strange - and not well documented.  Shouldn't happen for a Pod, but a VM running Hbone and multiple workloads ( with some custom implementation for delegation ) is reasonable. 

"Not support service" - why not ? I would think it's quite reasonable to define a Service with port name=hbone and on the well known port, and use it with workloads that are native implementations of Hbone. I understand that's not current implementation - but long term it seems very reasonable to define both hbone and the list of SA at service level, and potentially let CNI handle the LB ( as an option ).

Reply:
I suppose we could for service but having service vip in EDS should basically never happen?

Reply:
But we probably have the same in CDS where it would make more sense